substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/PosixJavaNetSubstitutions.java

```diff
@@ -94,17 +94,246 @@ private PosixJavaNetSubstitutions() {
+        //                                             "Bind failed");
+        //            } else {
+        //                NET_ThrowByNameWithLastError(env, JNU_JAVANETPKG "SocketException",
+        //                                             "Bind failed");
+        //            }
+        //            return;
+        //        }
+        if (JavaNetNetUtilMD.NET_Bind(fd, him, len_Pointer.read()) < 0) {
+            if (Errno.errno() == Errno.EADDRINUSE() || Errno.errno() == Errno.EADDRNOTAVAIL() || //
+                    Errno.errno() == Errno.EPERM() || Errno.errno() == Errno.EACCES()) {
+                throw new BindException(PosixUtils.lastErrorString("Bind failed"));
+            } else {
+                throw new SocketException(PosixUtils.lastErrorString("Bind failed"));
+            }
+        }
+        //
+        //    /* initialize the local port */
+        //        if (localport == 0) {
+        //        /* Now that we're a connected socket, let's extract the port number
+        //         * that the system chose for us and store it in the Socket object.
+        //         */
+        //            if (JVM_GetSockName(fd, (struct sockaddr *)&him, &len) == -1) {
+        //                NET_ThrowByNameWithLastError(env, JNU_JAVANETPKG "SocketException",
+        //                                             "Error getting socket name");
+        //                return;
+        //            }
+        //
+        //            localport = NET_GetPortFromSockaddr((struct sockaddr *)&him);
+        //
+        //            (*env)->SetIntField(env, this, pdsi_localPortID, localport);
+        //        } else {
+        //            (*env)->SetIntField(env, this, pdsi_localPortID, localport);
+        //        }
+        if (lport == 0) {
+            if (VmPrimsJVM.JVM_GetSockName(fd, him, len_Pointer) == -1) {
+                throw new SocketException(PosixUtils.lastErrorString("Error getting socket name"));
+            }
+            // 602         localport = NET_GetPortFromSockaddr((struct sockaddr *)&him);
+            lport = JavaNetNetUtilMD.NET_GetPortFromSockaddr(him);
+            // 603         (*env)->SetIntField(env, this, psi_localportID, localport);
+            Util_java_net_PlainDatagramSocketImpl.as_Target_java_net_DatagramSocketImpl(this).localPort = lport;
+        } else {
+            // 605         (*env)->SetIntField(env, this, psi_localportID, localport);
+            Util_java_net_PlainDatagramSocketImpl.as_Target_java_net_DatagramSocketImpl(this).localPort = lport;
+        }
     }
 
     @Substitute
     @SuppressWarnings({"static-method", "unused"})
-    protected void send(DatagramPacket p) throws IOException {
-        throw VMError.unsupportedFeature("Unimplemented: java.net.PlainDatagramSocketImpl.send(DatagramPacket)");
+    protected void send(DatagramPacket packet) throws IOException {
+        //    char BUF[MAX_BUFFER_LEN];
```

Comment:
Argh!  What happened to `BUF`?  Now I have to trace through the control flow to see if the uses are really replaced.

Oh, and what is worse: You have changed what might a (essentially free) stack allocation, probably in the common case, into (not free) calls to `malloc` and `free`.  At least it looks like you remembered to call `free` on it. :-)